From f399746ed8b403cd5f1a9f0b2e14a37f49ac625d Mon Sep 17 00:00:00 2001
From: charles <dminus@andestech.com>
Date: Wed, 24 Apr 2024 14:57:37 +0800
Subject: [PATCH 030/168] cache: andes: fix cpu_dcache_[wb/inval]_range()
 failed on 64-bit physical address

Bug 30361 - L2 cache operation does not support physical memory
larger than 4 GB.

This issue stemmed from using writel for writing 64-bit physical
addresses (PA) to the L2 CCTL access line register, as writel is
limited to 32-bit operations.
The solution involved replacing writel with writeq, enabling the
correct handling of 64-bit PAs.

Signed-off-by: charles <dminus@andestech.com>
---
 drivers/cache/andes_cache.c | 33 ++++++++++++++++++++++++++-------
 1 file changed, 26 insertions(+), 7 deletions(-)

diff --git a/drivers/cache/andes_cache.c b/drivers/cache/andes_cache.c
index 8b7b3abf1b33..2d77079c9b56 100644
--- a/drivers/cache/andes_cache.c
+++ b/drivers/cache/andes_cache.c
@@ -118,6 +118,30 @@ static inline uint32_t andes_cpu_l2c_get_cctl_status(void)
 	return readl(andes_priv.l2c_base + ANDES_L2C_REG_STATUS_OFFSET);
 }
 
+static void cpu_l2c_cctl(phys_addr_t pa, void __iomem *base,
+			 int mhartid, unsigned long ops)
+{
+#ifdef CONFIG_64BIT
+	writeq(pa, (base + ANDES_L2C_REG_CN_ACC_OFFSET(mhartid)));
+#else
+	/*
+	 * Considering RV32 potential to use over 4G memory,
+	 * the physical address is split into upper and lower 32 bits
+	 * and stored in a 64-bits PA-type L2C CCTL access line register.
+	 */
+	writel((pa & 0xFFFFFFFF),
+	       (base + ANDES_L2C_REG_CN_ACC_OFFSET(mhartid)));
+	writel((pa >> 32),
+	       (base + ANDES_L2C_REG_CN_ACC_OFFSET(mhartid) + 0x4));
+#endif /* !CONFIG_64BIT */
+
+	writel(ops, base + ANDES_L2C_REG_CN_CMD_OFFSET(mhartid));
+	while ((andes_cpu_l2c_get_cctl_status() &
+		ANDES_CCTL_L2_STATUS_CN_MASK(mhartid)) !=
+		ANDES_CCTL_L2_STATUS_IDLE)
+		;
+}
+
 static void andes_cpu_cache_operation(unsigned long start, unsigned long end,
 				       unsigned int l1_op, unsigned int l2_op)
 {
@@ -126,18 +150,13 @@ static void andes_cpu_cache_operation(unsigned long start, unsigned long end,
 	int mhartid = smp_processor_id();
 	unsigned long pa;
 
-	if (base) {
+	if (likely(base)) {
 		while (end > start) {
 			csr_write(ANDES_CCTL_REG_UCCTLBEGINADDR_NUM, start);
 			csr_write(ANDES_CCTL_REG_UCCTLCOMMAND_NUM, l1_op);
 
 			pa = virt_to_phys((void *)start);
-			writel(pa, base + ANDES_L2C_REG_CN_ACC_OFFSET(mhartid));
-			writel(l2_op, base + ANDES_L2C_REG_CN_CMD_OFFSET(mhartid));
-			while ((andes_cpu_l2c_get_cctl_status() &
-				ANDES_CCTL_L2_STATUS_CN_MASK(mhartid)) !=
-				ANDES_CCTL_L2_STATUS_IDLE)
-				;
+			cpu_l2c_cctl(pa, base, mhartid, l2_op);
 
 			start += line_size;
 		}
-- 
2.34.1

